python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,723 @@
+                # now we have a thumbnail on disk, either via the direct
+                # download, or via the url-fresh-then-download approach
+                # the file is downloaded with user-only permissions
+                # modify the permissions of the file so it's writeable by others
+                old_umask = os.umask(0)
+                try:
+                    os.chmod(thumb_path, 0666)
+                finally:
+                    os.umask(old_umask)
+
+        # finally, see if we should also load in the image
+        thumb_image = None
+        if thumb_path and os.path.exists(thumb_path):
+            if load_image:
+                # load the thumbnail into a QImage:
+                thumb_image = QtGui.QImage()
+                thumb_image.load(thumb_path)
+        else:
+            thumb_path = None
+
+        return {"action":"download_thumbnail", "thumb_path":thumb_path, "image":thumb_image}
+
+    def _on_task_completed(self, task_id, group, result):
+        """
+        Slot triggered when a task is completed.
+
+        :param task_id: The id of the task that has completed
+        :param group:   The group the task belongs to
+        :param result:  The task result
+        """
+        if group != self._bg_tasks_group:
+            # ignore - it isn't our task!
+            return
+
+        action = result.get("action")
+        if action in ["find", "create", "delete", "update"]:
+            self.work_completed.emit(str(task_id), action, {"sg":result["sg_result"]})
+        elif action == "schema":
+            self.work_completed.emit(str(task_id), "schema", {"fields":result["fields"], "types":result["types"]})
+        elif action == "method":
+            self.work_completed.emit(str(task_id), "method", {"return_value":result["result"]})
+        elif action == "check_thumbnail":
+            path = result.get("thumb_path", "")
+            if path:
+                # check found a thumbnail!
+                self.work_completed.emit(str(task_id), "find", {"thumb_path": path, "image":result["image"]})
+        elif action == "download_thumbnail":
+            # look up the primary thumbnail task id in the map:
+            thumb_task_id = self._thumb_task_id_map.get(task_id)
+            if thumb_task_id is not None:
+                del self._thumb_task_id_map[task_id]
+                self.work_completed.emit(str(thumb_task_id), 
+                                         "find", 
+                                         {"thumb_path": result["thumb_path"], "image":result["image"]})
+
+    def _on_task_failed(self, task_id, group, msg, tb):
+        """
+        Slot triggered when a task fails for some reason
+
+        :param task_id: The id of the task that failed
+        :param msg:     The error/exception message for the failed task
+        :param tb:      The stack trace of the exception raised by the failed task
+        """
+        if group != self._bg_tasks_group:
```

Comment:
actually, there is already a comment - the task doesn't belong to the task group (the task manager could be processing tasks from many different sources).